api/status.go

```diff
@@ -0,0 +1,142 @@
+	"github.com/mattermost/platform/utils"
+)
+
+func InitStatus() {
+	l4g.Debug(utils.T("api.status.init.debug"))
+
+	BaseRoutes.Users.Handle("/status", ApiUserRequiredActivity(getStatusesHttp, false)).Methods("GET")
+	BaseRoutes.WebSocket.Handle("get_statuses", ApiWebSocketHandler(getStatusesWebSocket))
+}
+
+func getStatusesHttp(c *Context, w http.ResponseWriter, r *http.Request) {
+	statusMap, err := GetAllStatuses()
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	w.Write([]byte(model.StringInterfaceToJson(statusMap)))
+}
+
+func getStatusesWebSocket(req *model.WebSocketRequest) (map[string]interface{}, *model.AppError) {
+	statusMap, err := GetAllStatuses()
+	if err != nil {
+		return nil, err
+	}
+
+	return statusMap, nil
+}
+
+func GetAllStatuses() (map[string]interface{}, *model.AppError) {
+	if result := <-Srv.Store.Status().GetOnlineAway(); result.Err != nil {
+		return nil, result.Err
+	} else {
+		statuses := result.Data.([]*model.Status)
+
+		statusMap := map[string]interface{}{}
+		for _, s := range statuses {
+			statusMap[s.UserId] = s.Status
+		}
+
+		return statusMap, nil
+	}
+}
+
+func SetStatusOnline(userId string, sessionId string) {
+	broadcast := false
+
+	var status *model.Status
+	var err *model.AppError
+	if status, err = GetStatus(userId); err != nil {
+		status = &model.Status{userId, model.STATUS_ONLINE, model.GetMillis()}
+		broadcast = true
+	} else {
+		if status.Status != model.STATUS_ONLINE {
+			broadcast = true
+		}
+		status.Status = model.STATUS_ONLINE
+		status.LastActivityAt = model.GetMillis()
+	}
+
+	statusCache.Add(status.UserId, status)
+
+	achan := Srv.Store.Session().UpdateLastActivityAt(sessionId, model.GetMillis())
+	schan := Srv.Store.Status().SaveOrUpdate(status)
+
+	if result := <-achan; result.Err != nil {
+		l4g.Error(utils.T("api.status.last_activity.error"), userId, sessionId, result.Err)
+	}
+
+	if result := <-schan; result.Err != nil {
+		l4g.Error(utils.T("api.status.save_status.error"), userId, result.Err)
+	}
+
+	if broadcast {
+		event := model.NewWebSocketEvent("", "", status.UserId, model.WEBSOCKET_EVENT_STATUS_CHANGE)
+		event.Add("status", model.STATUS_ONLINE)
+		go Publish(event)
+	}
+}
+
+func SetStatusOffline(userId string) {
+	status := &model.Status{userId, model.STATUS_OFFLINE, model.GetMillis()}
+
+	statusCache.Add(status.UserId, status)
+
+	if result := <-Srv.Store.Status().SaveOrUpdate(status); result.Err != nil {
+		l4g.Error(utils.T("api.status.save_status.error"), userId, result.Err)
+	}
+
+	event := model.NewWebSocketEvent("", "", status.UserId, model.WEBSOCKET_EVENT_STATUS_CHANGE)
+	event.Add("status", model.STATUS_OFFLINE)
+	go Publish(event)
+}
+
+func SetStatusAwayIfNeeded(userId string) {
+	status, err := GetStatus(userId)
+	if err != nil {
+		status = &model.Status{userId, model.STATUS_OFFLINE, 0}
+	}
+
+	if status.Status == model.STATUS_AWAY {
+		return
+	}
+
+	if model.GetMillis()-status.LastActivityAt <= *utils.Cfg.TeamSettings.UserStatusAwayTimeout*1000 {
```

Comment:
Can we move this logic to a function. Something like `isUserConsiteredAway(lastActivity int)`